File: src/main/java/model/PunchRecord.java
package model;

import java.sql.Date;
import java.sql.Time;
import java.text.MessageFormat;

@Model("punch_record")
public class PunchRecord {

    @Field("uid")
    @ForeignKey("id")
    @PrimaryKey
    private Person person;

    @Field("date")
    @PrimaryKey
    private Date date;

    @Field("time")
    private Time time;

    @Field("status")
    private Integer status;

    public PunchRecord() { }

    public PunchRecord(Person person, Date date, Time time, Integer status) {
        this.person = person;
        this.date = date;
        this.time = time;
        this.status = status;
    }

    public Person getPerson() {
        return person;
    }

    public void setPerson(Person person) {
        this.person = person;
    }

    public Date getDate() {
        return date;
    }

    public void setDate(Date date) {
        this.date = date;
    }

    public Time getTime() {
        return time;
    }

    public void setTime(Time time) {
        this.time = time;
    }

    public Status getStatus() {
        return new Status(status);
    }

    public void setStatus(Integer status) {
        this.status = status;
    }

    @Override
    public String toString() {
        return MessageFormat.format("PunchRecord'{'person={0}, date={1}, time={2}, status={3}'}'", person, date, time, new Status(status));
    }
}

File: src/main/java/model/PunchRecord.kt
package model

import java.sql.Date
import java.sql.Time
import java.text.MessageFormat
import java.text.SimpleDateFormat

@Model("punch_record")
class PunchRecord {
    @Field("uid")
    @PrimaryKey
    var person: String = ""

    @Field("date")
    @PrimaryKey
    var date: Date = Date(System.currentTimeMillis())

    @Field("time")
    var time: Time = Time(System.currentTimeMillis())

    @Field("status")
    private var status: Int = 0

    @Field("color")
    var color: String = CodeColor.RED.toString()

    constructor(person: Person, date: Date, time: Time, status: Int, color: String) {
        this.person = person.personId
        this.date = date
        this.time = time
        this.status = status
        this.color = color
    }

    constructor() {}
    constructor(person: Person, date: Date, time: Time, status: Int) {
        this.person = person.personId
        this.date = date
        this.time = time
        this.status = status
    }

    fun getStatus(): Status {
        return Status(status)
    }

    fun setStatus(status: Int) {
        this.status = status
    }

    fun getPerson() = SQL().query(Person::class.java, "uid" to person)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is PunchRecord) return false

        if (person != other.person) return false
        if (date != other.date) return false

        return true
    }

    override fun hashCode(): Int {
        var result = person.hashCode()
        result = 31 * result + date.hashCode()
        return result
    }

    override fun toString(): String {
        return "PunchRecord(person='$person', date=$date, time=$time, status=$status, color='$color')"
    }

}